### ISSUE TITLE
Confusion Regarding plot_contour

### ISSUE THREAD

**Issue by osama-usuf** [optuna/optuna#2524]
**Hello everybody,**

So I am using Optuna for optimizing a network with multiple hyper-parameters (5, to be exact). While I'm finding the visualization tools really helpful, I have a small confusion about the **`optuna.visualization.plot_contour`** method. When I use this method over my problem, Optuna provides a total of 25 plots (a separate contour plot for each possible pair of hyperparameters).

What I am confused by is the following: when using GridSampler, it is possible for two hyperparameters to have the same exact values over multiple trials, where the difference in each of these trials might be in one (or more) of the other three hyperparameters. Now, in a given contour plot (one of the 25) - how exactly is the objective value being calculated? Is it being summed/averaged over all trials, ignoring the other hyperparameters (this is what I believe is happening after going through the underlying code), or is it calculating partial dependencies in some other way? If this is the case, then doesn't it make more sense to have the contour plots depict the best objective values only (maximum/minimum, depending on the study direction)?

I'd appreciate any clarifications, thanks a bunch!

**Comment by fsmosca** [comment]
> What I am confused by is the following: when using GridSampler, it is possible for two hyperparameters to have the same exact values over multiple trials, where the difference in each of these trials might be in one (or more) of the other three hyperparameters. Now, in a given contour plot (one of the 25) - how exactly is the objective value being calculated?

Contour plot has 3 axis, the x, y and depth, like a bird's-eye view we are on top looking down. The x and y are the parameter values while the depth is the objective value shown in color because we are looking down we cannot see the depth difference but we can see which one is high or low by color. If x and y are the same for a given 2 trials on the same parameter pair for example then the location of the point is the same. 

For direction=maximize if the color around those points are darker the objective values are lower which means a bad parameter combination coming from two trials. If the colors are lighter that means the objective values of the 2 trials are both high - good. If the colors are in the middle not lighter and not darker that means one trial has a high objective value and one has a lower objective value. Perhaps a shade of blue to make it darker is in the formula in the calculation of objective color. As the objective value gets smaller, more shade is added.

**Comment by Crissman** [comment]
Thank you @fsmosca for the answer. 🙇  I believe this address this question, so closing this issue. Please feel free to re-open if the answer is not clear.

**Comment by BenjaminMidtvedt** [comment]
To be clear, if two trials have the same x and y values, is the color representative of the average objective value of those trials? A quick glance at the source suggested that the objective value of the last trial is used, but I may be mistaken.
